### ISSUE TITLE
Changes from MacOS not syncing to iOS

### ISSUE THREAD

**Issue by jcbwd** [laurent22/joplin#2382]
Host Computer:
MacBook Pro 2017, 13", running MacOS 10.15.2, and Joplin 1.0.178

Phone:
iPhone SE, 32GB, running iOS 13.3 and Joplin 10.0.43.

Steps to reproduce:
1. Create a library in Joplin for MacOS
2. Enable e2ee
3. Sync library via Dropbox
4. Install Joplin for iOS
5. Turn attachment downloading to "Automatic"
6. Sync library via Dropbox
7. Input master key to decrypt library
8. Decrypt library
9. Changes no longer sync between libraries. 

Logs from both iOS and MacOS are attached. MacOS log does not show anything remarkable. iOS log contains the following error:

"Processing a path that has already been done: 18e1f...md. sync_time was not updated? Remote item has an updated_time in the future?"

I assume this is an issue with the time stamp, which is breaking the sync, such that items changed in MacOS receive later timestamps than items changed on iOS? I have both devices set to update time automatically, so I'm not sure what might be causing the problem. Thanks for your help!

![IMG_0494](https://user-images.githubusercontent.com/3252189/73184156-e4fa1480-4113-11ea-9983-7a60dc01452b.PNG)
[log.txt](https://github.com/laurent22/joplin/files/4117641/log.txt)

**Comment by bedwardly-down** [comment]
> "Processing a path that has already been done: 18e1f...md. sync_time was not updated? Remote item has an updated_time in the future?"

This does seem like a timestamp issue. What is your time showing on both your phone and mac? Also, when you open the note in Joplin on your Mac, could you click on this [icon](https://fontawesome.com/v4.7.0/icon/info-circle) (Note Properties information) on your top toolbar and get the current Updated time from that section?

**Comment by jcbwd** [comment]
Absolutely!

As of right now, both devices are showing 12:40pm, EST. The timestamp on one of my new notes that isn't syncing says 12:39pm.

I did some digging, and found that Dropbox had incorrectly assigned me to GMT (+0hrs), when I should have been in EST (-5hrs). This meant that the modified times of my files in Dropbox showed in the future. I verified this with the note in question (even though it is encrypted), because the Dropbox homepage on the web places your most recently modified file first, making it possible to locate a recently changed encrypted note, in spite of its encrypted filename. After adjusting my timezone, that note now shows in Dropbox a modified timestamp of 12:40pm, 1 minute later than the timestamp in Joplin.

At this point it is 12:46, 6 minutes after the modified timestamp of the note in question. I ran a sync operation at 12:45 in Joplin for MacOS, followed by one in Joplin for iOS. The note in question is still not showing up in iOS, and I am getting the same error message on iOS.

I have a hunch that although Dropbox is now *displaying* the correct timestamp, it is not actually *assigning* the correct timestamp to the actual files on their server. Would that explain the issue?

**Comment by bedwardly-down** [comment]
I'm not sure beyond that, but resetting Joplin on each platform by reinstalling and resetting sync settings might solve the issue. Make sure whichever platform is correct is fully synced with Dropbox first

**Comment by jcbwd** [comment]
I fixed the timezone in Dropbox, backed up my notes, wiped Joplin completely from both devices, as well as my Dropbox account, and started over:
-Updated Joplin for MacOS to 179.
-Imported the notes into Joplin for MacOS.
-Synced to Dropbox.
-Synced to iOS. So far no problems.
-Turned on Encryption on MacOS.
-Synced MacOS.
-Synced iOS.
-Input Encryption Password for iOS.
-The problem is now repeating itself. I am seeing that there are a few different bugs related to Mac <> iOS sync issues: 2419, 2354. Any chance that these are somehow related, and that somehow MacOS and iOS aren't handling the timestamps right? At least on encrypted files?

**Comment by bedwardly-down** [comment]
That might be the case. I wish my mac hadn't kicked it because i would love to test this out to see what's what. Now, if you search around the issue tracker here, it seems like quite a few users are having issues with iOS 13.3 and OSX Catalina. Even I have had some minor issues here and there with my iPhone. Either way, the fact that quite a few people are having issues with the platforms means that it may be a bit difficult to figure out what are caused by Joplin and what are inherent to Apple's weird ecosystem quirks that pop up every so often. Ha

**Comment by wobblywob** [comment]
I have a similar issue. I'm a new user trying out Joplin. I've installed the client on macOS, Windows and iOS and notebooks and notes don't sync properly across devices.

Setup:
- OSes: macOS 10.15.4 (Joplin 1.0.201), Windows 10 1909 (Joplin 10.0.201), iOS 13.4 (Joplin 10.0.47);
- Backend: Nextcloud 17.0.3 (then updated to 18.0.3) on Debian 10;
- Timezones, dates and times are the same across devices;

Behavior:
Some notebooks synced when I've set it up, but new notebooks don't sync anymore. Some notes do sync even though their respective notebook don't.

I've deleted all apps and created a new joplin folder on nextcloud. The initial sync is successful and I can see 3 Welcome notebooks (2 x Desktop and 1 x Mobile) on all 3 devices. Then, new notebooks don't sync.

macOS and iOS have the following message in the log:

> 2020-04-21 18:58:54: "Processing a path that has already been done: 03639f50b85d4cc6af0fa78796f83974.md. sync_time was not updated? Remote item has an updated_time in the future?"

03639f50b85d4cc6af0fa78796f83974.md is Joplin's welcome message.

Windows logs don't show much:

> 2020-04-21 18:59:56: "Preparing scheduled sync"
> 2020-04-21 18:59:56: "Starting scheduled sync"
> 2020-04-21 18:59:58: "TaskQueue.stop: syncDownload: waiting for tasks to complete: 0"
> 2020-04-21 18:59:58: "TaskQueue.stop: syncDownload: Done, waited for 0"
> 2020-04-21 18:59:58: "BasicDelta: Report: {"timestamp":1587484721000,"older":30,"newer":0,"equal":1}"
> 2020-04-21 18:59:58: "Operations completed: "
> 2020-04-21 18:59:58: "fetchingTotal: -"
> 2020-04-21 18:59:58: "Total folders: 4"
> 2020-04-21 18:59:58: "Total notes: 18"
> 2020-04-21 18:59:58: "Total resources: 9"
> 2020-04-21 18:59:59: "DecryptionWorker: cannot start because no master key is currently loaded."

So, I've deleted the 3 welcome notebooks from the mac and synced windows. On Windows,  the new notebook created on the mac appeared, but the 3 deleted notebooks are still there. iOS didn't update anything.

Deleted the 3 welcome notebooks on windows. No change on Mac. 1 of them disappeared on iOS but its notes are in "conflicts".

Deleted the conflicting notes and the other 2 notebooks from iOS. Now windows synced the notebook created on iOS. MacOS still doesn't sync anything from the other 2 devices.

> 2020-04-21 19:11:10: "DecryptionWorker: cannot start because no master key is currently loaded."
2020-04-21 19:13:27: "Scheduling sync operation...", "0"
2020-04-21 19:13:27: "Preparing scheduled sync"
2020-04-21 19:13:27: "Starting scheduled sync"
2020-04-21 19:13:33: "Processing a path that has already been done: 51ca3b58752345968e8b880b7a8b6e18.md. sync_time was not updated? Remote item has an updated_time in the future?"
2020-04-21 19:13:33: "Operations completed: "
2020-04-21 19:13:33: "itemConflict: 3"
2020-04-21 19:13:33: "Total folders: 1"
2020-04-21 19:13:33: "Total notes: 1"
2020-04-21 19:13:33: "Total resources: 9"
2020-04-21 19:13:34: "DecryptionWorker: cannot start because no master key is currently loaded."

Created new notebook on MacOS. It appeared on Windows. Nothing in iOS.

So, it looks like macOS and iOS don't pull folders and sometimes don't properly push deletions, whereas Windows usually works as expected.

**Comment by wobblywob** [comment]
My bad. Windows was on a different time zone. I've checked the timestamp and indeed the .md has the timestamp in the future. 

What I still find a bit weird is that macOS and iOS wouldn't sync with one another even though they were creating correct stamps.

**Comment by jcbwd** [comment]
Ok, I tried something a bit different. To rule out a problem with Dropbox,
I set up a Nextcloud account on Hetzner. Then, I reinstalled Joplin on both
the Mac and the iPhone. Then tried syncing To Nextcloud. Looks like the
same result. That rules out a problem with Dropbox.

After some more digging, I did find one clue: Although the time stamp on
the files themselves is correct, the updated_time in the header of the
files is not.

For example: I created a note at 19:35, which synced at 19:37 according to
the log. However, when I opened that note in Nextcloud to check the header,
it says it was updated at 23:35, which is 4 hours in the future (see
attached).

How does the sync client interact with header? If either of the apps is
writing the wrong updated_time in the header, and the sync client is
checking the header, then perhaps we have our answer.
